/* DO NOT MODIFY: this is automatically generated by the cpptypes */

#include <cstring>
#include <stdexcept>
#include <cstdint>

#ifdef _WIN32
#define PYAPI __declspec(dllexport)
#else
#define PYAPI
#endif

static char* copy_error_message(const char* original) {
    auto n = std::strlen(original);
    auto copy = new char[n + 1];
    std::strcpy(copy, original);
    return copy;
}

void extract_r_missing_double(double*);

uint8_t fetch_csv_booleans(void*, int32_t, uint8_t*);

uint8_t fetch_csv_numbers(void*, int32_t, double*, uint8_t*);

void fetch_csv_strings(void*, int32_t, char*);

void free_csv(void*);

void get_csv_column_stats(void*, int32_t, int32_t*, int32_t*, int32_t*);

int32_t get_csv_num_fields(void*);

int32_t get_csv_num_records(void*);

uint8_t get_csv_string_stats(void*, int32_t, int32_t*, uint8_t*);

void* load_csv(const char*);

void* load_list_hdf5(const char*, const char*, int32_t);

void* load_list_json(const char*, int32_t);

void uzuki2_free_list(void*);

int32_t uzuki2_get_boolean_vector_length(void*);

void uzuki2_get_boolean_vector_mask(void*, uint8_t*);

uint8_t uzuki2_get_boolean_vector_values(void*, uint8_t*);

int32_t uzuki2_get_external_index(void*);

int32_t uzuki2_get_integer_vector_length(void*);

void uzuki2_get_integer_vector_mask(void*, uint8_t*);

uint8_t uzuki2_get_integer_vector_values(void*, int32_t*);

void* uzuki2_get_list_element(void*, int32_t);

int32_t uzuki2_get_list_length(void*);

uint8_t uzuki2_get_list_named(void*);

void uzuki2_get_list_names_contents(void*, char*);

uint64_t uzuki2_get_list_names_lengths(void*, int32_t*);

int32_t uzuki2_get_node_type(void*);

int32_t uzuki2_get_number_vector_length(void*);

void uzuki2_get_number_vector_mask(void*, uint8_t*);

uint8_t uzuki2_get_number_vector_values(void*, double*);

void* uzuki2_get_parent_node(void*);

uint8_t uzuki2_get_string_vector_contents(void*, char*);

int32_t uzuki2_get_string_vector_length(void*);

uint64_t uzuki2_get_string_vector_lengths(void*, int32_t*);

void uzuki2_get_string_vector_mask(void*, uint8_t*);

void validate_csv(const char*);

void validate_list_hdf5(const char*, const char*, int32_t);

void validate_list_json(const char*, int32_t);

extern "C" {

PYAPI void free_error_message(char** msg) {
    delete [] *msg;
}

PYAPI void py_extract_r_missing_double(double* buffer, int32_t* errcode, char** errmsg) {
    try {
        extract_r_missing_double(buffer);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI uint8_t py_fetch_csv_booleans(void* ptr, int32_t column, uint8_t* contents, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = fetch_csv_booleans(ptr, column, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI uint8_t py_fetch_csv_numbers(void* ptr, int32_t column, double* contents, uint8_t* mask, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = fetch_csv_numbers(ptr, column, contents, mask);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_fetch_csv_strings(void* ptr, int32_t column, char* contents, int32_t* errcode, char** errmsg) {
    try {
        fetch_csv_strings(ptr, column, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI void py_free_csv(void* ptr, int32_t* errcode, char** errmsg) {
    try {
        free_csv(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI void py_get_csv_column_stats(void* ptr, int32_t column, int32_t* type, int32_t* size, int32_t* loaded, int32_t* errcode, char** errmsg) {
    try {
        get_csv_column_stats(ptr, column, type, size, loaded);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI int32_t py_get_csv_num_fields(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = get_csv_num_fields(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_get_csv_num_records(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = get_csv_num_records(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI uint8_t py_get_csv_string_stats(void* ptr, int32_t column, int32_t* lengths, uint8_t* mask, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = get_csv_string_stats(ptr, column, lengths, mask);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void* py_load_csv(const char* path, int32_t* errcode, char** errmsg) {
    void* output = NULL;
    try {
        output = load_csv(path);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void* py_load_list_hdf5(const char* path, const char* name, int32_t n, int32_t* errcode, char** errmsg) {
    void* output = NULL;
    try {
        output = load_list_hdf5(path, name, n);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void* py_load_list_json(const char* path, int32_t n, int32_t* errcode, char** errmsg) {
    void* output = NULL;
    try {
        output = load_list_json(path, n);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_uzuki2_free_list(void* ptr, int32_t* errcode, char** errmsg) {
    try {
        uzuki2_free_list(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI int32_t py_uzuki2_get_boolean_vector_length(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_boolean_vector_length(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_uzuki2_get_boolean_vector_mask(void* ptr, uint8_t* mask, int32_t* errcode, char** errmsg) {
    try {
        uzuki2_get_boolean_vector_mask(ptr, mask);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI uint8_t py_uzuki2_get_boolean_vector_values(void* ptr, uint8_t* contents, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = uzuki2_get_boolean_vector_values(ptr, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_uzuki2_get_external_index(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_external_index(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_uzuki2_get_integer_vector_length(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_integer_vector_length(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_uzuki2_get_integer_vector_mask(void* ptr, uint8_t* mask, int32_t* errcode, char** errmsg) {
    try {
        uzuki2_get_integer_vector_mask(ptr, mask);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI uint8_t py_uzuki2_get_integer_vector_values(void* ptr, int32_t* contents, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = uzuki2_get_integer_vector_values(ptr, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void* py_uzuki2_get_list_element(void* ptr, int32_t i, int32_t* errcode, char** errmsg) {
    void* output = NULL;
    try {
        output = uzuki2_get_list_element(ptr, i);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_uzuki2_get_list_length(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_list_length(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI uint8_t py_uzuki2_get_list_named(void* ptr, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = uzuki2_get_list_named(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_uzuki2_get_list_names_contents(void* ptr, char* contents, int32_t* errcode, char** errmsg) {
    try {
        uzuki2_get_list_names_contents(ptr, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI uint64_t py_uzuki2_get_list_names_lengths(void* ptr, int32_t* lengths, int32_t* errcode, char** errmsg) {
    uint64_t output = 0;
    try {
        output = uzuki2_get_list_names_lengths(ptr, lengths);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_uzuki2_get_node_type(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_node_type(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_uzuki2_get_number_vector_length(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_number_vector_length(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_uzuki2_get_number_vector_mask(void* ptr, uint8_t* mask, int32_t* errcode, char** errmsg) {
    try {
        uzuki2_get_number_vector_mask(ptr, mask);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI uint8_t py_uzuki2_get_number_vector_values(void* ptr, double* contents, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = uzuki2_get_number_vector_values(ptr, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void* py_uzuki2_get_parent_node(void* ptr, int32_t* errcode, char** errmsg) {
    void* output = NULL;
    try {
        output = uzuki2_get_parent_node(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI uint8_t py_uzuki2_get_string_vector_contents(void* ptr, char* contents, int32_t* errcode, char** errmsg) {
    uint8_t output = 0;
    try {
        output = uzuki2_get_string_vector_contents(ptr, contents);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI int32_t py_uzuki2_get_string_vector_length(void* ptr, int32_t* errcode, char** errmsg) {
    int32_t output = 0;
    try {
        output = uzuki2_get_string_vector_length(ptr);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI uint64_t py_uzuki2_get_string_vector_lengths(void* ptr, int32_t* lengths, int32_t* errcode, char** errmsg) {
    uint64_t output = 0;
    try {
        output = uzuki2_get_string_vector_lengths(ptr, lengths);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
    return output;
}

PYAPI void py_uzuki2_get_string_vector_mask(void* ptr, uint8_t* mask, int32_t* errcode, char** errmsg) {
    try {
        uzuki2_get_string_vector_mask(ptr, mask);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI void py_validate_csv(const char* path, int32_t* errcode, char** errmsg) {
    try {
        validate_csv(path);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI void py_validate_list_hdf5(const char* path, const char* name, int32_t n, int32_t* errcode, char** errmsg) {
    try {
        validate_list_hdf5(path, name, n);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

PYAPI void py_validate_list_json(const char* path, int32_t n, int32_t* errcode, char** errmsg) {
    try {
        validate_list_json(path, n);
    } catch(std::exception& e) {
        *errcode = 1;
        *errmsg = copy_error_message(e.what());
    } catch(...) {
        *errcode = 1;
        *errmsg = copy_error_message("unknown C++ exception");
    }
}

}
